### DISCUSSION TITLE
collection onSnapshot document spread error

### DISCUSSION THREAD

**Discussion by puchesjr** [firebase/firebase-js-sdk#4806] [Q&A] [answered]
I am encountering an error when attempting to use v9 as follows - perhaps I am not doing something properly, but the error message is rather cryptic.

```ts
onSnapshot(
      collection(firestore, colRef),
      (snap) => {
        if (snap.empty) {
          setStatus('fetched');
          return;
        }
        const documents = snap.docs.map((doc) => ({
          id: doc.id,
          ...doc.data() // error: Spread types may only be created from object types
        }));
        setData(documents);
        setStatus('fetched');
      },
      (error) => {
        setData([error]);
        setStatus('error');
      }
    );
```

**Comment by dbanisimov** [comment]
This looks like a typescript error for me. Is there a chance that your `colRef` is incorrectly typed causing `doc.data()` to be something that is not an object?

**Reply by puchesjr** [reply]
colRef is a string which should work properly.  Perhaps it cannot take an unknown string?

**Reply by dbanisimov** [reply]
Yeah `colRef` should be string in this case.

`collection(firestore, colRef)` is of type `CollectionReference<DocumentData>` by default, so I would expect that `onSnapshot` should infer `QuerySnapshot<DocumentData>` type for `(snap) => {}` argument, but for some reason it fails to do that and infers `QuerySnapshot<unknown>` instead. That causes the spread error. Maybe that's something that Firebase team can take a look at.

From my personal experience, I would advise to use[ `.withConverter()`](https://modularfirebase.web.app/reference/firestore_.collectionreference.withconverter) method. It just makes data conversions, like you're trying to do, very easy and works well with types inference throughout.

**Reply by puchesjr** [reply]
I don't think `.withConverter()` should matter here.  I do see the point of using `.withConverter()` if I was attempting to map data to a Class as specified in the docs, but I simply want the entire document and the id which is the reason for my destructured/spread of the `doc.data()`

**Reply by puchesjr** [reply]
this works, but the above does not. as you can see the above uses the spread operator and this example does not.
```
onSnapshot(
      collection(firestore, colRef),
      (snap) => {
        if (snap.empty) {
          setStatus('fetched');
          return;
        }
        const documents = snap.docs.map((doc) => ({
          id: doc.id,
          data: doc.data()
        }));
        setData(documents);
        setStatus('fetched');
      },
      (error) => {
        setData([error]);
        setStatus('error');
      }
    );
```

**Comment by samtstern** [comment, marked as answer]
@puchesjr I am able to reproduce this and since it's a bug I am going to move it to an issue.

**Reply by samtstern** [reply]
https://github.com/firebase/firebase-js-sdk/issues/4812
